### PR TITLE
Update Marker

### PR DESCRIPTION
**Pull request by NaamaHolzer**
- Let the user update an existing marker
- Use enum to perform all marker back-end operations in a single servlet

**Note:**
Right now updating a marker changes the data stored in the DB but doesn't show it on the web page until it reloads. This will be fixed in the next PR when the info window becomes a separate component.

### CODE REVIEW

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 1]
```diff
@@ -73,6 +100,20 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     /** Stores a marker in Datastore. */
     public static void storeMarker(Marker marker) {
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        datastore.put(Marker.toEntity(marker));
+        Entity markerEntity = new Entity("Marker");
+        datastore.put(Marker.toEntity(marker, markerEntity));
     }
-}
+
+    public static void updateMarker(Marker marker) throws EntityNotFoundException {
```
Comment (NoyFlat) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:107]: Forgot description of the function

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 2]
```diff
@@ -73,6 +100,20 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Key markerEntityKey = KeyFactory.createKey("Marker", marker.getId());
+        Entity markerEntity = datastore.get(markerEntityKey);
+        datastore.put(Marker.toEntity(marker, markerEntity));
+    }
+
+    public static void deleteMarker(long markersId) {
```
Comment (NoyFlat) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:114]: Forgot description of the function

**Review by NoyFlat — APPROVED** [review]

File: capstone-project/src/main/angular-webapp/src/app/map/map.component.ts [hunk 3]
```diff
@@ -79,16 +85,16 @@ export class MapComponent implements OnInit {
       return containerDiv;
     }
 
-    // Sends a marker to the backend for saving.
-    function postMarker(marker) {
+    // Performs a back end action on a marker - create / update / delete.
```
Comment (orarch) [inline capstone-project/src/main/angular-webapp/src/app/map/map.component.ts:88]: super-nit: backend

File: capstone-project/src/main/java/com/google/sps/data/Marker.java [hunk 4]
```diff
@@ -125,9 +125,9 @@ public static Marker fromEntity(Entity entity){
                 .build();
     }
 
-    /** Creates a marker entity from a Marker */
-    public static Entity toEntity(Marker marker){
-        Entity markerEntity = new Entity("Marker");
+    // Copies the fields of a Marker class instance to a Marker entity.
+    public static Entity toEntity(Marker marker, Entity markerEntity){
+
```
Comment (orarch) [inline capstone-project/src/main/java/com/google/sps/data/Marker.java:130]: I liked it better when the creation of the class was here. 
Why did you change it?

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 5]
```diff
@@ -73,6 +100,22 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     /** Stores a marker in Datastore. */
     public static void storeMarker(Marker marker) {
```
Comment (orarch) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:101]: make this method private

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 6]
```diff
@@ -73,6 +100,22 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        datastore.put(Marker.toEntity(marker));
+        Entity markerEntity = new Entity("Marker");
+        datastore.put(Marker.toEntity(marker, markerEntity));
     }
-}
+
+    /** Updates an existing marker's data */
+    public static void updateMarker(Marker marker) throws EntityNotFoundException {
```
Comment (orarch) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:108]: private

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 7]
```diff
@@ -73,6 +100,22 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Key markerEntityKey = KeyFactory.createKey("Marker", marker.getId());
+        Entity markerEntity = datastore.get(markerEntityKey);
+        datastore.put(Marker.toEntity(marker, markerEntity));
+    }
+
+    /** Deletes a marker */
+    public static void deleteMarker(long markersId) {
```
Comment (orarch) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:116]: private

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 8]
```diff
@@ -73,6 +100,22 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+        Entity markerEntity = new Entity("Marker");
+        datastore.put(Marker.toEntity(marker, markerEntity));
     }
-}
+
+    /** Updates an existing marker's data */
+    public static void updateMarker(Marker marker) throws EntityNotFoundException {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Key markerEntityKey = KeyFactory.createKey("Marker", marker.getId());
+        Entity markerEntity = datastore.get(markerEntityKey);
```
Comment (orarch) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:111]: Why do we need to read the entity first? Don't we have all the updated data to store?

Comment (NaamaHolzer) [reply capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:111]: In order to overwrite an existing Marker entity and not create a new one - I first found the relevant marker entity and then updated its properties. This way the entity's ID doesn't change and only the data fields are updated when we call the put() function. 
Did you mean that it'd be better if we sent the key to toEntity() and find the corresponding entity there instead of in this function?

File: capstone-project/src/main/java/com/google/sps/data/Marker.java (hunk 4 above)
Comment (NaamaHolzer) [reply capstone-project/src/main/java/com/google/sps/data/Marker.java:130]: I wanted to distinguish between two use cases of this method - 1. Storing a new marker which means creating a new entity 2. Updating an existing entity. When the creation of the class was here - this function created a new entity each time it was called. Now I can create a new entity in the storeMarker method and retrieve one in the updateMarker method (as explained in the comment below), and then pass it to set the properties here, in the toEntity method. 
Is there a better alternative?

Comment (orarch) [reply capstone-project/src/main/java/com/google/sps/data/Marker.java:130]: You can overload methods, having:

public static Entity toEntity(Marker marker);
public static Entity toEntity(Marker marker, Entity entity);

I suggest to better document that too.

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java (hunk 8 above)
Comment (orarch) [reply capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:111]: That's possible. 

You can have toEntity method that accepts ID as an argument (overloading the existing one.

**Review by orarch — APPROVED** [review]
